pkg/arvo/app/chat-hook.hoon

```diff
@@ -0,0 +1,366 @@
+    !!
+  =/  box=(unit mailbox)  (chat-scry pax)
+  ?~  box  !!
+  :_  this
+  [ost.bol %diff %chat-update [%create (slav %p i.pax) pax]]~
+::
+++  diff-permission-update
+  |=  [wir=wire diff=permission-update]
+  ^-  (quip move _this)
+  ~&  %permission-update
+  :_  this
+  ?-  -.diff
+      %create  ~
+      %delete  ~
+      %add     (handle-permissions [%add path.diff who.diff])
+      %remove  (handle-permissions [%remove path.diff who.diff])
+  ==
+::
+++  handle-permissions
+  |=  [kind=?(%add %remove) pax=path who=(set ship)]
+  ^-  (list move)
+  ?>  ?=([* *] pax)
+  ?.  =(%chat i.pax)  ~
+  ::  check path to see if this is a %read permission
+  ?.  =(%read (snag (dec (lent pax)) `(list @t)`pax))
+    ~
+  =/  sup
+    %-  ~(gas by *(map [ship path] bone))
+    %+  turn  ~(tap by sup.bol)
+    |=([=bone anchor=[ship path]] [anchor bone])
+  %-  zing
+  %+  turn  ~(tap in who)
+  |=  check-ship=ship
+  ?:  (permitted-scry [(scot %p check-ship) pax])
+    ::  if ship is permitted to be subscribed, do nothing
+    ~
+  :: if ship is not permitted, quit their subscription
+  =/  mail-path
+    (oust [(dec (lent t.pax)) (lent t.pax)] `(list @t)`t.pax)
+  =/  bne  (~(get by sup) [check-ship [%mailbox mail-path]])
+  ?~(bne ~ [u.bne %quit ~]~)
+::
+++  diff-chat-update
+  |=  [wir=wire diff=chat-update]
+  ^-  (quip move _this)
+  ?:  (team:title our.bol src.bol)
+    (handle-local diff)
+  (handle-foreign diff)
+::
+++  handle-local
+  |=  diff=chat-update
+  ^-  (quip move _this)
+  ?-  -.diff
+      %keys    [~ this]
+      %config  [~ this]
+      %create  [~ this]
+      %read    [~ this]
+      %delete
+    ?.  (~(has by synced) path.diff)
+      [~ this]
+    :_  this(synced (~(del by synced) path.diff))
+    :-  (chat-poke diff)
+    [ost.bol %pull [%mailbox path.diff] [our.bol %chat-store] ~]~
+  ::
+      %message
+    :_  this
+    %+  turn  (prey:pubsub:userlib [%mailbox path.diff] bol)
+    |=  [=bone *]
+    ^-  move
+    [bone %diff [%chat-update diff]]
+  ::
+  ==
+::
+++  handle-foreign
+  |=  diff=chat-update
+  ^-  (quip move _this)
+  ?-  -.diff
+      %keys    [~ this]
+      %config  [~ this]
+      %read    [~ this]
+      %create
+    :_  this
+    ?>  ?=([* ^] path.diff)
+    =/  shp  (~(get by synced) path.diff)
+    ?~  shp  ~
+    ?.  =(u.shp src.bol)  ~
+    [(chat-poke [%create ship.diff t.path.diff])]~
+  ::
+      %delete
+    ?>  ?=([* ^] path.diff)
+    =/  shp  (~(get by synced) path.diff)
+    ?~  shp
+      [~ this]
+    ?.  =(u.shp src.bol)
+      [~ this]
+    :_  this(synced (~(del by synced) path.diff))
+    :-  (chat-poke diff)
+    [ost.bol %pull [%mailbox path.diff] [src.bol %chat-hook] ~]~
+  ::
+      %message
+    :_  this
+    ?>  ?=([* ^] path.diff)
+    =/  shp  (~(get by synced) path.diff)
+    ?~  shp  ~
+    ?.  =(src.bol u.shp)  ~
+    [(chat-poke diff)]~
+  ::
+  ==
+::
+++  quit
+  |=  wir=wire
+  ^-  (quip move _this)
+  ~&  chat-hook-quit+wir
```

Comment:
Should remove ~&'s

Reply:
These are valuable input for the user, as these cases are rare and only really initiated by semantic actions on other ships (a chatroom being deleted, or a user being banned).